File: ai/select-algorithm-dotnet/src/IvfDemo.cs
/// IVF (Inverted File) vector index for Azure DocumentDB.
/// Best for: Datasets with fewer than 10,000 documents.
/// Cluster tier: M10 or higher.
/// Key parameters: numLists (cluster count).

namespace SelectAlgorithm;

using MongoDB.Driver;
using MongoDB.Bson;

public static class IvfDemo
{
    public static void CreateIvfIndex(IMongoCollection<BsonDocument> collection, string vectorField, int dimensions, string similarity, int numLists = 10)
    {
        Console.WriteLine($"Creating IVF vector index on field '{vectorField}'...");

        Utils.DropVectorIndexes(collection, vectorField);

        var command = new BsonDocument
        {
            { "createIndexes", collection.CollectionNamespace.CollectionName },
            { "indexes", new BsonArray
                {
                    new BsonDocument
                    {
                        { "name", $"ivf_index_{vectorField}" },
                        { "key", new BsonDocument(vectorField, "cosmosSearch") },
                        { "cosmosSearchOptions", new BsonDocument
                            {
                                { "kind", "vector-ivf" },
                                { "dimensions", dimensions },
                                { "similarity", similarity },
                                { "numLists", numLists }
                            }
                        }
                    }
                }
            }
        };

        collection.Database.RunCommand<BsonDocument>(command);
        Console.WriteLine("IVF vector index created successfully");
    }

    public static void Run()
    {
        Console.WriteLine(new string('=', 60));
        Console.WriteLine("  IVF Vector Index - Select Algorithm Demo");
        Console.WriteLine("  Best for: < 10,000 documents");
        Console.WriteLine(new string('=', 60));

        var databaseName = Environment.GetEnvironmentVariable("AZURE_DOCUMENTDB_DATABASENAME") ?? "Hotels";
        var dataFile = Environment.GetEnvironmentVariable("DATA_FILE_WITH_VECTORS") ?? "../../data/Hotels_Vector.json";
        var vectorField = Environment.GetEnvironmentVariable("EMBEDDED_FIELD") ?? "DescriptionVector";
        var model = Environment.GetEnvironmentVariable("AZURE_OPENAI_EMBEDDING_MODEL") ?? "text-embedding-3-small";
        var dimensions = int.Parse(Environment.GetEnvironmentVariable("EMBEDDING_DIMENSIONS") ?? "1536");
        var batchSize = int.Parse(Environment.GetEnvironmentVariable("LOAD_SIZE_BATCH") ?? "100");
        var similarity = Environment.GetEnvironmentVariable("SIMILARITY") ?? "COS";

        var mongoClient = Utils.GetMongoClientPasswordless();
        var embeddingClient = Utils.GetEmbeddingClient();

        try
        {
            var database = mongoClient.GetDatabase(databaseName);
            var collection = database.GetCollection<BsonDocument>("hotels_ivf");

            var data = Utils.ReadJsonFile(dataFile);
            var documents = data.Where(d => d.Contains(vectorField)).ToList();
            Console.WriteLine($"\nLoaded {documents.Count} documents with embeddings");

            Utils.InsertData(collection, documents, batchSize);

            CreateIvfIndex(collection, vectorField, dimensions, similarity);
            Console.WriteLine("Waiting for index to build...");
            Thread.Sleep(3000);

            var query = "quintessential lodging near running trails, eateries, retail";
            var results = Utils.PerformVectorSearch(collection, embeddingClient, query, vectorField, model);
            Utils.PrintSearchResults(results, "IVF");
        }
        finally
        {
            mongoClient.Cluster.Dispose();
        }
    }
}

File: ai/select-algorithm-java/src/main/java/com/azure/documentdb/selectalgorithm/IvfDemo.java
package com.azure.documentdb.selectalgorithm;

import com.azure.ai.openai.OpenAIClient;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import org.bson.Document;

import java.util.List;

public class IvfDemo {

    private static final String COLLECTION_NAME = "hotels_ivf";
    private static final String QUERY = "quintessential lodging near running trails, eateries, retail";

    public static void createIvfIndex(MongoCollection<Document> collection, String vectorField, int dimensions, String similarity) {
        System.out.println("  Creating IVF vector index...");

        Document indexDefinition = new Document()
                .append("name", "ivf_index_" + vectorField)
                .append("key", new Document(vectorField, "cosmosSearch"))
                .append("cosmosSearchOptions", new Document()
                        .append("kind", "vector-ivf")
                        .append("dimensions", dimensions)
                        .append("similarity", similarity)
                        .append("numLists", 10));

        Document command = new Document("createIndexes", collection.getNamespace().getCollectionName())
                .append("indexes", List.of(indexDefinition));

        collection.getDatabase().runCommand(command);
        System.out.println("  IVF index created successfully.");
    }

    public static void run() {
        System.out.println("\n========================================");
        System.out.println("  IVF (Inverted File) Index Demo");
        System.out.println("========================================\n");

        String databaseName = Utils.getEnv("AZURE_DOCUMENTDB_DATABASENAME", "Hotels");
        String dataFile = Utils.getEnv("DATA_FILE_WITH_VECTORS", "../data/Hotels_Vector.json");
        String vectorField = Utils.getEnv("EMBEDDED_FIELD", "contentVector");
        int dimensions = Integer.parseInt(Utils.getEnv("EMBEDDING_DIMENSIONS", "1536"));
        String similarity = Utils.getEnv("SIMILARITY", "COS");
        String model = Utils.getEnv("AZURE_OPENAI_EMBEDDING_MODEL", "text-embedding-3-small");

        try (MongoClient mongoClient = Utils.getMongoClient()) {
            MongoDatabase database = mongoClient.getDatabase(databaseName);
            MongoCollection<Document> collection = database.getCollection(COLLECTION_NAME);

            // Load and insert data
            System.out.println("  Loading data from: " + dataFile);
            List<Document> data = Utils.readJsonFile(dataFile);
            System.out.printf("  Loaded %d documents%n", data.size());

            // Drop existing collection to start fresh
            collection.drop();
            System.out.println("  Collection reset.");

            Utils.insertData(collection, data, 100);

            // Create IVF index
            createIvfIndex(collection, vectorField, dimensions, similarity);

            // Perform vector search
            OpenAIClient aiClient = Utils.getOpenAIClient();
            System.out.println("\n  Performing vector search with IVF index...");
            List<Document> results = Utils.performVectorSearch(
                    collection, aiClient, QUERY, vectorField, model, 5);

            Utils.printResults(results);
        }

        System.out.println("  IVF Demo complete.\n");
    }
}
